velox/exec/Spiller.cpp

```diff
@@ -0,0 +1,402 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/Spiller.h"
+
+#include "velox/common/base/AsyncSource.h"
+
+#include <folly/ScopeGuard.h>
+
+namespace facebook::velox::exec {
+
+void Spiller::extractSpill(folly::Range<char**> rows, RowVectorPtr* resultPtr) {
+  if (!*resultPtr) {
+    *resultPtr = std::static_pointer_cast<RowVector>(
+        BaseVector::create(rowType_, rows.size(), &spillPool()));
+  } else {
+    (*resultPtr)->prepareForReuse();
+    (*resultPtr)->resize(rows.size());
+  }
+  auto result = resultPtr->get();
+  auto& types = container_.columnTypes();
+  for (auto i = 0; i < types.size(); ++i) {
+    container_.extractColumn(rows.data(), rows.size(), i, result->childAt(i));
+  }
+  auto& aggregates = container_.aggregates();
+  auto numKeys = types.size();
+  for (auto i = 0; i < aggregates.size(); ++i) {
+    aggregates[i]->finalize(rows.data(), rows.size());
```

Comment:
We have discussed that Aggregate::finalize should not allocate memory. Have we made that change already? I don't see the comments for finalize to reflect this requirement.

```
  // Finalizes the state in groups. Defaults to no op for cases like
  // sum and max.
  virtual void finalize(char** groups, int32_t numGroups) = 0;
```